pkg/client/unversioned/remotecommand/remotecommand.go

```diff
@@ -21,141 +21,118 @@ import (
-	"k8s.io/kubernetes/pkg/runtime"
 	"k8s.io/kubernetes/pkg/util"
 	"k8s.io/kubernetes/pkg/util/httpstream"
 	"k8s.io/kubernetes/pkg/util/httpstream/spdy"
 )
 
-type upgrader interface {
-	upgrade(*client.Request, *client.Config) (httpstream.Connection, error)
+type Executor interface {
+	Stream(stdin io.Reader, stdout, stderr io.Writer, tty bool) error
 }
 
-type defaultUpgrader struct{}
-
-func (u *defaultUpgrader) upgrade(req *client.Request, config *client.Config) (httpstream.Connection, error) {
-	return req.Upgrade(config, spdy.NewRoundTripper)
+type StreamExecutor interface {
+	Executor
+	httpstream.Dialer
 }
 
-type Streamer struct {
-	req    *client.Request
-	config *client.Config
-	stdin  io.Reader
-	stdout io.Writer
-	stderr io.Writer
-	tty    bool
+type streamExecutor struct {
+	upgrader  httpstream.UpgradeRoundTripper
+	transport http.RoundTripper
 
-	upgrader upgrader
+	method string
+	url    *url.URL
 }
 
-// Executor executes a command on a pod container
-type Executor struct {
-	Streamer
-	command []string
-}
+// NewExecutor connects to the provided server and upgrades the connection to
+// multiplexed bidirectional streams. The current implementation uses SPDY,
+// but this could be replaced with HTTP/2 once it's available, or something else.
+// TODO: the common code between this and portforward could be abstracted.
+func NewExecutor(config *client.Config, method string, url *url.URL) (StreamExecutor, error) {
+	tlsConfig, err := client.TLSConfigFor(config)
+	if err != nil {
+		return nil, err
+	}
 
-// New creates a new RemoteCommandExecutor
-func New(req *client.Request, config *client.Config, command []string, stdin io.Reader, stdout, stderr io.Writer, tty bool) *Executor {
-	return &Executor{
-		command: command,
-		Streamer: Streamer{
-			req:    req,
-			config: config,
-			stdin:  stdin,
-			stdout: stdout,
-			stderr: stderr,
-			tty:    tty,
-		},
+	upgradeRoundTripper := spdy.NewRoundTripper(tlsConfig)
+	wrapper, err := client.HTTPWrappersForConfig(config, upgradeRoundTripper)
+	if err != nil {
+		return nil, err
 	}
-}
 
-type Attach struct {
-	Streamer
+	return &streamExecutor{
+		upgrader:  upgradeRoundTripper,
+		transport: wrapper,
+		method:    method,
+		url:       url,
+	}, nil
 }
 
-// NewAttach creates a new RemoteAttach
-func NewAttach(req *client.Request, config *client.Config, stdin io.Reader, stdout, stderr io.Writer, tty bool) *Attach {
-	return &Attach{
-		Streamer: Streamer{
-			req:    req,
-			config: config,
-			stdin:  stdin,
-			stdout: stdout,
-			stderr: stderr,
-			tty:    tty,
-		},
+// NewStreamExecutor upgrades the request so that it supports multiplexed bidirectional
+// streams. This method takes a stream upgrader and an optional function that is invoked
+// to wrap the connection.
+func NewStreamExecutor(upgrader httpstream.UpgradeRoundTripper, fn func(http.RoundTripper) http.RoundTripper, method string, url *url.URL) (StreamExecutor, error) {
```

Comment:
Wrapping the round tripper.  I'll add a test case to clarify it shouldn't
be removed.

On Oct 7, 2015, at 3:16 PM, Andy Goldstein <notifications@github.com> wrote:

In pkg/client/unversioned/remotecommand/remotecommand.go
<https://github.com/kubernetes/kubernetes/pull/14618#discussion_r41433112>:

>  }
>
> -// NewAttach creates a new RemoteAttach
> -func NewAttach(req *client.Request, config *client.Config, stdin io.Reader, stdout, stderr io.Writer, tty bool) *Attach {
> -	return &Attach{
> -		Streamer: Streamer{
> -			req:    req,
> -			config: config,
> -			stdin:  stdin,
> -			stdout: stdout,
> -			stderr: stderr,
> -			tty:    tty,
> -		},
> +// NewStreamExecutor upgrades the request so that it supports multiplexed bidirectional
> +// streams. This method takes a stream upgrader and an optional function that is invoked
> +// to wrap the connection.
> +func NewStreamExecutor(upgrader httpstream.UpgradeRoundTripper, fn func(http.RoundTripper) http.RoundTripper, method string, url *url.URL) (StreamExecutor, error) {

Ok. The usage in OpenShift doesn't pass a fn... What was your plan for that?

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/14618/files#r41433112>.